File: 0238-product-of-array-except-self/0238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        // int[] answer = new int[nums.length];
       
        
        int[] ans = new int[nums.length];
        ans[0] = 1;
        int right = 1;
        
        for (int i = 1; i < nums.length; i++) {
            ans[i] = nums[i-1] * ans[i-1];
        }
        
        for (int i = nums.length - 1; i>=0; i--) {
            ans[i] = ans[i] * right;
            right *= nums[i];
        }
        
        return ans;
    }
}

File: 0238-product-of-array-except-self/0238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1] * len(nums)
        left = 1
        
        for i in range(0, len(nums)):
            res[i] = left
            left *= nums[i]
            

        right = 1
        for i in range(len(nums) - 1, -1, -1):
            res[i] *= right
            right *= nums[i]
            
        
        return res
        
        
        
#         left = [1] * len(nums)
#         right = [1] * len(nums)
        
#         for i in range(1, len(nums)):
#             left[i] = left[i - 1] * nums[i - 1]
        
        
#         for i in range(len(nums) - 2, -1, -1):
#             right[i] = right[i + 1] * nums[i + 1]
        
        
#         for i in range(len(left)):
#             right[i] *= left[i]
        
#         return right